translation/message_format.rst

```diff
@@ -0,0 +1,436 @@
+.. index::
+    single: Translation; Message Format
+
+How to translate messages using the ICU MessageFormat
+=====================================================
+
+.. versionadded:: 4.2
+
+   Support for ICU MessageFormat was introduced in Symfony 4.2.
+
+Messages (i.e. strings) in applications are almost never completely static.
+They contain variables or other complexer logic like pluralization. In order to
+handle this, the Translator component supports the `ICU MessageFormat`_ syntax.
+
+Using the ICU Message Format
+----------------------------
+
+In order to use the ICU Message Format, the :ref:`message domain
+<using-message-domains>` has to be suffixed with ``intl-icu``:
+
+======================  ===============================
+Normal file name        ICU Message Format filename
+======================  ===============================
+``messages.en.yml``     ``messages+intl-icu.en.yml``
+``messages.fr_FR.xlf``  ``messages+intl-icu.fr_FR.xlf``
+``admin.en.yml``        ``admin+intl-icu.en.yml``
+======================  ===============================
+
+All messages in this file will now be processed by the
+:phpclass:`MessageFormatter` during translation.
+
+.. _component-translation-placeholders:
+
+Message Placeholders
+--------------------
+
+The basic usage of the MessageFormat allows you to use placeholder (called
+*arguments* in ICU MessageFormat) in your messages:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        # translations/messages.en.yaml
+        say_hello: 'Hello {name}!'
+
+    .. code-block:: xml
+
+        <!-- translations/messages.en.xlf -->
+        <?xml version="1.0"?>
+        <xliff version="1.2" xmlns="urn:oasis:names:tc:xliff:document:1.2">
+            <file source-language="en" datatype="plaintext" original="file.ext">
+                <body>
+                    <trans-unit id="say_hello">
+                        <source>say_hello</source>
+                        <target>Hello {name}!</target>
+                    </trans-unit>
+                </body>
+            </file>
+        </xliff>
+
+    .. code-block:: php
+
+        // translations/messages.en.php
+        return [
+            'say_hello' => "Hello {name}!",
+        ];
+
+Everything within the curly braces (``{...}``) is processed by the formatter
+and replaced by it's placeholder::
+
+    // ...
+    echo $translator->trans('say_hello', ['name' => 'Fabien']); // Hello Fabien!
```

Comment:
Let's align the comments at the end of these two lines to improve readability a bit.